### PR TITLE
Use custom hcc/HIP, purge hcSPARSE

### PR DESCRIPTION
**Pull request by iotamudelta**
* purge hcSPARSE now that rocSPARSE is available
* integrate a custom hcc and HIP
* hcc brings two important compiler fixes (fixes hundreds of unit tests)
* HIP brings a smart dispatcher that allows us to avoid a lot of static_casts (we haven't yet removed the automatic static_casts but this catches some occurrences the script did not catch)
* mark 5 unit tests skipping that have regressed w/ the new hcc (we don't know yet what is at fault)
* optimize bitonic sort - the comparator is always an empty struct - therefore passing it by value saves at least 3 bytes. It also removes an ambiguity around passing references to `__global__` functions

### CODE REVIEW

**Comment by iotamudelta** [comment]
@ezyang @Jorghi12 please respin the docker image after this - let's discuss how we can get the deb's to you offline.

**Review by ezyang — APPROVED** [review]

File: aten/src/THC/THCSortUtils.cuh [hunk 1]
```diff
@@ -142,7 +142,7 @@ bitonicSortKVInPlace(TensorInfo<K, IndexType> keys,
                      IndexType keySliceStride,
                      TensorInfo<V, IndexType> values,
                      IndexType valueSliceStride,
-                     const Comparator& comp) {
+                     const Comparator comp) {
```
Comment (ezyang) [inline aten/src/THC/THCSortUtils.cuh]: This change makes me nervous.

1. There are a lot of other `const Comparator&` arguments in this file; why is only this one changed?

2. What if Comparator is not cheap to copy? Now you've hugely pessimized this function.

**Comment by iotamudelta** [comment]
random failure 

@pytorchbot retest this please

**Comment by iotamudelta** [comment]
download failure...

@pytorchbot retest this please

**Comment by iotamudelta** [comment]
@ezyang any more concerns here? :-)

**Comment by ezyang** [comment]
Just needs someone to say that they are going to respin the docker image.

**Comment by iotamudelta** [comment]
@ezyang awesome, thanks!

@Jorghi12 could you do this? I'll make you a package of the deb's necessary tomorrow!

**Pull request closed without merging**